tls/s2n_ktls_io.c

```diff
@@ -268,6 +269,46 @@ S2N_RESULT s2n_ktls_recvmsg(void *io_context, uint8_t *record_type, uint8_t *buf
     return S2N_RESULT_OK;
 }
 
+/* ktls does not support updating keys, so we should kill the connection when the
+ * key encryption limit is reached. We could get the actual sequence number from
+ * the kernel with getsockopt, but that requires a surprisingly expensive syscall.
+ *
+ * Instead, we track the estimated sequence number and enforce the limit based
+ * on that estimate.
+ */
+static S2N_RESULT s2n_ktls_enforce_estimated_record_limit(
+        struct s2n_connection *conn, size_t bytes_written)
+{
+    RESULT_ENSURE_REF(conn);
+    if (bytes_written == 0) {
+        return S2N_RESULT_OK;
+    }
+
+    struct s2n_blob seq_num = { 0 };
+    RESULT_GUARD(s2n_connection_get_sequence_number(conn, conn->mode, &seq_num));
```

Comment:
Here's where I don't remember how this works and get confused. In KTLS we don't keep track of the sequence number right? Where does KTLS actually increment our stored sequence number?

Reply:
We increment the sequence number on line 295 :)

With ktls we don't keep track of the sequence number. The kernel handles sending records, so we don't even know how many records it chooses to send. But since we need to track key limits for safety, we're incrementing the sequence number here with an estimate of how many records were sent.

Reply:
Sorry, obviously I was very confused when I first read this PR 😄